### ISSUE TITLE
[Bug]: Large variance in leading silence before speech in Qwen3-TTS  x-vector-only mode

### ISSUE THREAD

**Issue by iancarrasco-b10** [vllm-project/vllm-omni#4966]
### Environment
**Deployment**
- Docker image: `vllm/vllm-omni:v0.24.0` (container `qwen3tts`), serving `:8091`
- Server reported version: 0.24.0
**vLLM / vLLM-Omni**
- vLLM 0.24.0
- vLLM-Omni 0.24.0
**CUDA / PyTorch**
- PyTorch 2.11.0+cu130 (CUDA build 13.0), CUDA available: yes
- torchaudio 2.11.0+cu130
- GPU: NVIDIA H100 80GB HBM3
**Python libraries**
- transformers 5.12.1
- numpy 2.2.6
- triton 3.6.0
- flashinfer-python 0.6.12
**OS**
- Ubuntu 22.04.5 LTS (container)

### Your code version

vllm-omni 0.24.0

### 🐛 Describe the bug

**Summary**
In `x_vector_only_mode=True` (Base voice clone), generated audio starts with a
substantial, highly variable leading silence — mean ~326 ms, ranging 125–460 ms
across prompts. This inflates perceived time-to-first-audio and makes a fixed
server-side trim hard to tune.

**Environment**
- Qwen3-TTS Base, single-H100 deploy (vllm-omni), 24 kHz output
- Endpoint: `POST /v1/audio/speech`, `stream=false` (non-streaming path, so this
  reflects the model's generated codes, not the streaming/chunk pipeline)
- Reference: `jim.wav` (ref audio's own leading silence is ~20 ms)

**Measurement**
8 prompts × 5 reps = 40 samples. Leading silence = time until audio first crosses
a −34 dB (peak 0.02) energy threshold, 5 ms frame resolution.

| Prompt | leading silence (ms) |
|--------|----------------------|
| Hello there, how are you doing today? | 330 |
| Absolutely, let me check that for you right now. | 340 |
| The quick brown fox jumps over the lazy dog. | 125 |
| Seven. | 380 |
| Thanks for calling, my name is Alex and I can help. | 460 |
| Could you confirm your account number, please? | 375 |
| It looks like your order shipped this morning. | 205 |
| Let's schedule a follow up for next Tuesday. | 390 |

Overall: mean 326 ms · median 358 ms · stdev 101.6 ms · range 125–460 ms.

**Notes**
- Within-prompt stdev is 0 (deterministic; `seed=42`). The variance is entirely
  cross-prompt / content-driven, not run-to-run noise.
- Confirmed on the non-streaming path, so the silence is baked into the generated
  codec frames (not a streaming/chunking artifact).
- For contrast, ICL mode (ref codec prefix + ref_text) shows a much tighter onset,
  since generation continues from in-context speech rather than a cold start.

**Impact**
- Perceived TTFA is worse than the transport TTFA metric suggests (first audio
  bytes are silent).
- A single fixed leading-silence trim threshold must cope with a 125–460 ms
  moving target; the shortest cases risk clipping soft onsets (e.g. fricatives).

**Possible directions**
1. Server-side leading-silence trim on the emit path (cosmetic; robust; mode-agnostic).
2. Short in-context codec prefix ("mini-ICL") in x-vector mode to prime phonation
   while keeping the x-vector timbre anchor — reuses existing ICL prefill +
   code2wav context-trim machinery.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by IneshReddy249** [comment]
I'd like to take this one. Really clear write-up.
@iancarrasco-b10 quick question before I start — would you prefer the server-side silence trim, or the mini-ICL prefix? I'd lean toward the trim first (detect where speech actually starts and cut the leading silence, carefully so soft sounds don't get clipped). I can test it on an H100/H200 with your 40-sample setup.
Happy to do the other approach if you'd rather.

**Comment by iancarrasco-b10** [comment]
I'm a bit hesistant about the trimming approach as we don't want to unintentionally trim quiet speech as silence. I feel like it would introduce another silence threshold hyperparameter that would need tuning. This may very well be a structural issue in the model itself. So would be great if we can get more eyes to discuss the high level issue before converging on implementation

**Comment by IneshReddy249** [comment]
Makes sense — I'd rather not add a fragile silence threshold either. Happy to dig into the root cause first: I can run your 40 samples and log where speech actually starts in each one, to see if the silence is consistent per input or changes with a fixed seed. That'll tell us if it's a structural cold-start issue and whether priming is the better fix. I'll hold off on writing code until we've settled the direction.

**Comment by IneshReddy249** [comment]
> Spent a while on this. Short version: the silence is the talker sampling silence tokens in the opening frames, and everything else follows from that.
>
> **It's the codes, not Code2Wav.** Encoded real speech trimmed to start loud at sample 0, decoded it back: frame-0 peak 0.397 → 0.330, no ramp. Then spliced the talker's first generated frames into the middle of a loud sequence — still decode quiet (0.046 / 0.010 vs ~0.2 either side). Those codes are quiet wherever you put them. Encoding digital silence gives cb0 `[1995, 215, 212, ...]`, and `1995` is what every generation opens with.
>
> **The variance is the sampler.** Your stdev=0 is real, but your 40 samples were 5 reps at one fixed seed — each prompt replayed one trajectory. Vary the seed and a single prompt swings 330–740 ms, wider than your whole 125–460 ms cross-prompt range. Per-frame peaks, same prompt, 5 seeds: 0, 3, 1, 4, 1 leading quiet frames. At `temperature=0.9, top_k=50` the draw decides how much silence you get.
>
> **Underneath that, a content effect.** Set `temperature: 0.0` and the variance goes to zero, which exposes it:
>
> ```
> 610 ms   Could you confirm your account number, please?
> 670 ms   Could you help me?
> 590 ms   Can you confirm your account number?
> 585 ms   Can you repeat that?
> ---
>  70 ms   Would you confirm your account number?
> 150 ms   Could I confirm your account number?
>  65 ms   Please confirm your account number.
>  60 ms   Hello. Could you confirm your account number?
> ```
>
> Sentence-initial "Could you" / "Can you" → ~600 ms, deterministic, first-frame peaks of 0.0016 (nothing, not soft speech). Not "Would you", not "Could I". Any prefix kills it. At the token level both triggers open `[1995, 215, 1521, 690, ...]` — silence tokens — while the controls open with speech codes.
>
> So it's one mechanism: the talker draws silence tokens early. The variance is *how many* it draws; the 600 ms prompts are the ones where it draws them nearly every time.
>
> Don't reach for greedy as a fix — at `temperature: 0.0`, "Can you hear me?" generates a **326-second** clip of junk. Filing separately.
>
> Fix in the next comment.
>
> *(v0.24.0, default deploy, H200. Also: `seed` is silently dropped when decode batch > 1 under full cudagraphs — 8 concurrent requests at seed=42 → 6 different outputs. Matches the TODO at `qwen3_tts_talker.py:328-350`. Separate issue.)*

**Comment by iancarrasco-b10** [comment]
Great investigation Inesh! Based on this being rooted in the Talker do we potentially need to do something like logit biasing to discourage the model from outputting initial silence tokens, do you have any thoughts here?

**Comment by iancarrasco-b10** [comment]
Im wondering if this is also inherent even with the vanilla HF transformers implementation

**Comment by IneshReddy249** [comment]
Yes — and it works better than I expected. It fixes the variance too, not just the "Could you" cases.

To your second question first: **this is inherent to vanilla HF.** Everything below is in-process `qwen_tts` with HF `generate`, no vllm-omni involved.

Built the silence vocab by encoding digital silence and collecting codebook-0 tokens — `[212, 215, 462, 619, 690, 763, 951, 1181, 1225, 1318, 1657, 1995]` — then a step-aware `LogitsProcessor` that masks them for the **first 3 decode frames only**. Sampling (temp 0.9, top_k 50), 10 seeds per prompt, median/range in ms:

```
                                             ban=0        ban=3
TRIG  Could you help me?                   107 / 595    70 / 140
TRIG  Can you repeat that?                 372 / 865    67 /  95
TRIG  Could you confirm your account...    107 / 595    70 / 140
CTRL  Would you help me?                   110 / 385    75 /  75
CTRL  Please confirm your account number.  252 / 570    70 /  95
CTRL  Thanks for calling, my name is Alex   70 / 500    50 / 100
CTRL  Seven.                                70 / 470    50 /  80
```

Range collapses on **every** row, controls included. So the variance *was* the silence tokens — the model rolls dice on silence-vs-speech in the opening frames, and how many it draws is the spread. Ban them for 3 frames and the dice stop mattering. Three is enough: ban=5 and ban=8 give byte-identical output. Whisper transcripts are clean, durations drop by exactly the silence removed, no clipping, no runaway.

One detail I liked: the banned generation opens with token `1221` — the same token `"Would you help me?"` opens with unbanned. The model always had a good speech onset available; the silence mode was just winning the argmax.

**The catch is plumbing.** In vllm-omni the talker's layer-0 token comes from vLLM's sampler, not HF `generate`, so this needs a step-aware logits processor on the engine side. My prototype monkeypatches `talker.generate`, which obviously isn't shippable. Two questions before I try:

1. Does vllm-omni support per-request logits processors with decode-step awareness, or is that new plumbing?
2. I'd derive the silence vocab at init (encode zeros → collect cb0) rather than hardcode it, since it's checkpoint-specific. Any objection?

Separately, #5041 — smaller and orthogonal: the *reference's* own leading silence propagates into the output (405 ms-lead ref → 270 ms onset; trimmed → 131 ms). Input-side, doesn't touch any of the above.

**Comment by IneshReddy249** [comment]
Answered my own questions — turns out no new plumbing is needed.

`gpu_ar_model_runner` already passes `sampling_metadata` into `compute_logits`, and the talker already accepts the argument and ignores it. The decode step is just `len(sampling_metadata.output_token_ids[i])` — I probed it and it grows 0, 1, 2, 3 per frame as you'd hope. `compute_logits` also runs outside the cudagraph, so a plain Python branch is fine there.

So it's ~40 lines in `qwen3_tts_talker.py`, same `masked_fill` pattern as the existing `_codec_disallowed_mask`. The silence vocab is derived from the checkpoint at load time by encoding silence, not hardcoded.

Numbers hold on the serve path, not just in-process. Default deploy, 10 seeds per prompt, median/range in ms:

```
                                             ban=0        ban=3
TRIG  Could you help me?                   107 / 595    60 /  65
TRIG  Can you repeat that?                 372 / 865    50 /  65
CTRL  Would you help me?                   110 / 385    70 /  45
CTRL  Please confirm your account number.  252 / 570    55 /  95
CTRL  Thanks for calling, my name is Alex   70 / 500    55 /  65
CTRL  Seven.                                70 / 470    25 /  95
```

Transcripts clean, durations drop by exactly the silence removed, ICL unaffected, holds at batch=8, no runaway.

PR: #5048

One thing I'd value your read on, and I flagged it in the PR: if a request gets preempted and resumed, does `output_token_ids` still reflect the true decode step? If it resets, the ban would re-fire mid-utterance and suppress a pause the model actually wanted. I couldn't force a preemption to test it.

**Comment by iancarrasco-b10** [comment]
I'd think we'd only want this suppression to be on the first decode for a brand new request, I'm not sure if we can determine if a request is a continuation of a previous one

**Comment by IneshReddy249** [comment]
Good question — I went and read the code rather than guess.

The step index comes from `len(sampling_metadata.output_token_ids[i])`, and in `gpu_input_batch.py:349` that's a reference to the `Request` object's live list:

```python
self.req_output_token_ids[req_index] = request.output_token_ids
```

So it's the request's generated-token *history*, not a counter I'm keeping. And `Scheduler._preempt_request` (`scheduler.py:1107`) only sets `request.status = RequestStatus.PREEMPTED` and frees KV blocks — it never clears `output_token_ids`, since the engine needs those to re-prefill and continue.

A request preempted at frame 20 therefore comes back with `len(oti) == 20`, and `20 < 3` is False. The ban is already inert. A counter would have exactly the bug you're describing; a history can't.

To be clear about what I actually verified: this is from reading the code, not from watching a preemption happen. I tried to force one with a 15% KV budget and 48 concurrent long requests and couldn't get the scheduler to preempt. If you have a config that reliably triggers it I'm happy to run it.

The gap I'd worry about more is **streaming** — I haven't tested `stream=true` at all. If the streaming path builds `sampling_metadata` differently, the step count might not carry across chunks. Worth checking before this lands.

**Comment by IneshReddy249** [comment]
> Tested streaming — the ban holds there too.
>
> `ban=3`, 6 seeds per prompt, median / range in ms:
>
> ```
>                                         non-streaming    streaming
> Could you help me?                        52 /  25       57 /  65
> Can you repeat that?                      42 /  65       50 /  70
> Would you help me?                        70 /  15       75 / 110
> Thanks for calling, my name is Alex...    50 /  55       65 /  75
> ```
>
> Same band either way, nowhere near the 385–865 ms unbanned spread. So `output_token_ids` carries the decode step across streaming chunks the same as in the non-streaming path.
>
> That was the gap I flagged, so it's closed. PR description updated — #5048.

**Comment by iancarrasco-b10** [comment]
Yeah streaming should still be fine as requests are split on sentence boundaries so it pretty much resolves to the non-streaming case under the hood
